diff --git a/.travis.yml b/.travis.yml
deleted file mode 100644
index 859da98..0000000
-- a/.travis.yml
+++ /dev/null
@@ -1,46 +0,0 @@
sudo: required
services:
 - docker

env:
 - FAST_API_APP_NAME=calculator-api
 - FAST_API_APP_NAME_TEST=calculator-api-test

stages:
 - name: tests
 - name: checks
 - name: build
   if: branch = master

before_install:
 - docker build -t $DOCKER_USERNAME/$FAST_API_APP_NAME_TEST -f Dockerfile.dev .
 - docker build -t $DOCKER_USERNAME/$FAST_API_APP_NAME .


jobs:
 include:
  - stage: tests
    name: running_unit_tests
    script:
     - docker run $DOCKER_USERNAME/$FAST_API_APP_NAME_TEST /bin/bash -c "poe test"

  - stage: checks
    name: synchronized_hooks
    script:
     - docker run $DOCKER_USERNAME/$FAST_API_APP_NAME_TEST /bin/bash -c "git init && git add . && poe checks"

  - stage: build
    script:
      - docker build -t $FAST_API_APP_NAME .
      - echo "$DOCKER_PASSWORD" | docker login -u $DOCKER_USERNAME --password-stdin
      - docker tag $FAST_API_APP_NAME $DOCKER_USERNAME/$FAST_API_APP_NAME:latest
      - docker push $DOCKER_USERNAME/$FAST_API_APP_NAME:latest

deploy:
- provider: script
- script:
  - ssh $DIGITAL_OCEAN_USER@$DROPLET_IP
  - docker pull $DOCKER_USERNAME/$FAST_API_APP_NAME:latest
  - docker run -d -p 8080:8080 $DOCKER_USERNAME/$FAST_API_APP_NAME:latest
- on:
    branch: master
