diff --git a/inlineplz/interfaces/github.py b/inlineplz/interfaces/github.py
index 45d4a329..3feeeb2f 100644
--- a/inlineplz/interfaces/github.py
++ b/inlineplz/interfaces/github.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 from __future__ import absolute_import
 from __future__ import division
from __future__ import print_function
 from __future__ import unicode_literals
 
 import random
@@ -39,10 +40,9 @@ def __init__(self, owner, repo, pr, token, url=None):
     def pr_commits(pull_request):
         # github3 has naming/compatibility issues
         try:
-            commits = [c for c in pull_request.commits()]
            return [c for c in pull_request.commits()]
         except (AttributeError, TypeError):
-            commits = [c for c in pull_request.iter_commits()]
-        return commits
            return [c for c in pull_request.iter_commits()]
 
     def out_of_date(self):
         """Check if our local latest sha matches the remote latest sha"""
@@ -89,6 +89,7 @@ def post_messages(self, messages, max_comments):
                     messages_posted += 1
                     if max_comments >= 0 and messages_posted > max_comments:
                         break
        print('{} messages posted to Github.'.format(messages_to_post))
         return messages_to_post
 
     def is_duplicate(self, message, position):
