diff --git a/lisp/magit-wip.el b/lisp/magit-wip.el
index ff7b69c..e5df97c 100644
--- a/lisp/magit-wip.el
++ b/lisp/magit-wip.el
@@ -1,6 +1,6 @@
 ;;; magit-wip.el --- commit snapshots to work-in-progress refs  -*- lexical-binding: t -*-
 
-;; Copyright (C) 2010-2017  The Magit Project Contributors
;; Copyright (C) 2010-2018  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
 ;; lists all contributors.  If not, see http://magit.vc/authors.
