diff --git a/.github/workflows/go.yml b/.github/workflows/go.yml
new file mode 100644
index 0000000..9b43014
--- /dev/null
++ b/.github/workflows/go.yml
@@ -0,0 +1,22 @@
# This file adheres to the YAML5 style.
{
  "name": "Go",
  "on": ["push", "pull_request"],
  "jobs": {
    "build": {
      "name": "Build",
      "runs-on": "ubuntu-latest",
      "steps": [
        {
          "name": "Set up Go 1.18",
          "uses": "actions/setup-go@v1",
          "with": {"go-version": 1.18},
          "id": "go",
        },
        {"name": "Check out code into the Go module directory", "uses": "actions/checkout@v1"},
        {"name": "Lint", "run": "make lint"},
        {"name": "Test", "run": "make test"},
      ],
    },
  },
}
