diff --git a/scripts/travis-test-app.sh b/scripts/test-app.sh
similarity index 73%
rename from scripts/travis-test-app.sh
rename to scripts/test-app.sh
index eef2f216..3e48ccb4 100755
-- a/scripts/travis-test-app.sh
+++ b/scripts/test-app.sh
@@ -6,6 +6,6 @@ tag=$(cat .dockertag)
 
 docker run \
   --rm \
  -e LOCAL_SETTINGS=settings_local_travis \
+  -e LOCAL_SETTINGS=settings_local_ci \
   $repo:$tag \
   ./scripts/test.sh
