diff --git a/scripts/travis-build-image.sh b/scripts/build-image.sh
similarity index 79%
rename from scripts/travis-build-image.sh
rename to scripts/build-image.sh
index 2e878615..d309a208 100755
-- a/scripts/travis-build-image.sh
+++ b/scripts/build-image.sh
@@ -1,10 +1,10 @@
#!/bin/sh
+#!/bin/bash
 set -eu
 
 docker --version
 
 repo="cybernetisk/in-backend"
tag=$(date -u +%Y%m%d-%H%M)-$TRAVIS_BUILD_NUMBER
+tag=$(date -u +%Y%m%d-%H%M)-${GITHUB_SHA:0:8}
 echo $repo >.dockerrepo
 echo $tag >.dockertag
 
