diff --git a/.github/dependabot.yaml b/.github/dependabot.yaml
new file mode 100644
index 0000000..83b642c
--- /dev/null
++ b/.github/dependabot.yaml
@@ -0,0 +1,20 @@
version: 2
updates:
  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "weekly"
      day: "sunday"
      time: "16:00"
    groups:
      github-actions:
        patterns: [ "*" ]
  - package-ecosystem: "pip"
    directory: "/"
    schedule:
      interval: "weekly"
      day: "sunday"
      time: "16:00"
    groups:
      python-requirements:
        patterns: [ "*" ]
