diff --git a/.github/actions/build_ci/README.md b/.github/actions/build_ci/README.md
new file mode 100644
index 0000000..3b0d531
--- /dev/null
++ b/.github/actions/build_ci/README.md
@@ -0,0 +1,21 @@
# Libmetal build check docker action

This action test builds for a specified target.

## Inputs

### `target`

**Required** the build target. Default `"linux"`.
The supported taget are:
  linux
  generic arm
  freertos
  zephyr


## Example usage

uses: ./.github/actions/build_ci
with:
  target: freertos
\ No newline at end of file
