diff --git a/.travis/linux.script.sh b/.travis/linux.script.sh
index 03a7cd5ee01..4944f7e894e 100755
-- a/.travis/linux.script.sh
+++ b/.travis/linux.script.sh
@@ -1,10 +1,14 @@
#!/usr/bin/env sh
+#!/usr/bin/env bash
 # Author: Ivan Mincik, ivan.mincik@gmail.com
 
 set -e
 
 export CC="ccache $CC"
 export CXX="ccache $CXX"
+export MAKEFLAGS="-j $(nproc) --no-keep-going"
+
+echo "MAKEFLAGS is '$MAKEFLAGS'"
+
 ./configure --host=x86_64-linux-gnu \
             --build=x86_64-linux-gnu \
             --prefix=/usr/lib \
@@ -28,7 +32,7 @@ export CXX="ccache $CXX"
             --with-freetype-includes=/usr/include/freetype2/ \
             --with-postgres-includes=/usr/include/postgresql/ \
             --with-proj-share=/usr/share/proj \
            --with-python \
            --with-cairo
+            --with-cairo \
+            --with-pdal
 
make -j2
+make CFLAGS="$CFLAGS $GRASS_EXTRA_CFLAGS" CXXFLAGS="$CXXFLAGS $GRASS_EXTRA_CXXFLAGS"
