diff --git a/.github/workflows/main.yml b/.github/workflows/main.yml
new file mode 100644
index 0000000..0ceae65
--- /dev/null
++ b/.github/workflows/main.yml
@@ -0,0 +1,15 @@
on:
  push: { branches: [ master ] }
  pull_request: { branches: [ master ] }

jobs:

  test:
    runs-on: ubuntu-latest
    steps:
      - run: corepack enable
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with: { node-version: '18.x', cache: yarn }
      - run: yarn install --immutable
      - run: yarn test --ci
