diff --git a/.gitmodules b/.gitmodules
index 8da5fd7..ccb6efe 100644
--- a/.gitmodules
++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "MumbleKit"]
 	path = MumbleKit
-	url = http://github.com/mumble-voip/mumblekit.git
	url = https://github.com/mumble-voip/mumblekit.git
 [submodule "Dependencies/fmdb"]
 	path = Dependencies/fmdb
-	url = http://github.com/ccgus/fmdb.git
	url = https://github.com/ccgus/fmdb.git
