diff --git a/.travis/before_script/linux_clang_debug_static.sh b/.travis/before_script/linux_clang_debug_static.sh
deleted file mode 100755
index 619cedc8..00000000
-- a/.travis/before_script/linux_clang_debug_static.sh
+++ /dev/null
@@ -1,7 +0,0 @@
#!/usr/bin/env bash

cmake -H. -B${BUILD_DIR} \
    -DCMAKE_BUILD_TYPE=Debug \
    -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain/Clang.cmake \
    -DUJIT_TEST_LIB_TYPE=static \
    -DUJIT_TEST_OPTIONS="-O4"
