diff --git a/.travis/before_script/linux_clang_release_shared.sh b/.travis/before_script/linux_clang_release_shared.sh
deleted file mode 100755
index d22fe362..00000000
-- a/.travis/before_script/linux_clang_release_shared.sh
+++ /dev/null
@@ -1,7 +0,0 @@
#!/usr/bin/env bash

cmake -H. -B${BUILD_DIR} \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain/Clang.cmake \
    -DUJIT_TEST_LIB_TYPE=shared \
    -DUJIT_TEST_OPTIONS="-O4"
