diff --git a/ruby/travis-test.sh b/ruby/travis-test.sh
index 75db7d931f..cbe7cd983c 100755
-- a/ruby/travis-test.sh
+++ b/ruby/travis-test.sh
@@ -5,20 +5,24 @@ set -e
 
 test_version() {
   version=$1
  if [ "$version" == "jruby" ] ; then
+  if [ "$version" == "jruby-1.7" ] ; then
     # No conformance tests yet -- JRuby is too broken to run them.
     bash --login -c \
      "rvm install $version && rvm use $version && \
+      "rvm install $version && rvm use $version && rvm get head && \
        which ruby && \
+       git clean -f && \
        gem install bundler && bundle && \
        rake test"
   else
     bash --login -c \
       "rvm install $version && rvm use $version && \
        which ruby && \
+       git clean -f && \
        gem install bundler && bundle && \
        rake test &&
       cd ../conformance && make test_ruby"
+       rake gc_test &&
+       cd ../conformance && make test_ruby &&
+       cd ../ruby/compatibility_tests/v3.0.0 && ./test.sh"
   fi
 }
 
