diff --git a/csharp/.gitignore b/csharp/.gitignore
index 07ea90ac40..8ba8849985 100644
--- a/csharp/.gitignore
++ b/csharp/.gitignore
@@ -1,16 +1,10 @@
-#
-# 	Untracked directories
-#
-src/AddressBook/bin
-src/AddressBook/obj
-src/Google.Protobuf/bin/
-src/Google.Protobuf/obj/
-src/Google.Protobuf.Conformance/bin/
-src/Google.Protobuf.Conformance/obj/
-src/Google.Protobuf.Test/bin/
-src/Google.Protobuf.Test/obj/
-src/Google.Protobuf.JsonDump/bin/
-src/Google.Protobuf.JsonDump/obj/
# Output
bin
obj
project.lock.json
TestResult.xml

# Possibly legacy now?
 mono/bin
 mono/tmp
 mono/protoc
@@ -23,6 +17,7 @@ lib/NUnit
 #
 # 	Untracked files
 #
.vs
 *.user
 *.suo
 *.nupkg
@@ -33,3 +28,4 @@ mono/.libs
 mono/*.exe
 mono/*.dll
 lib/protoc.exe
*.ncrunch*
