diff --git a/lisp/magit-obsolete.el b/lisp/magit-obsolete.el
index b9a3003..b5439cb 100644
--- a/lisp/magit-obsolete.el
++ b/lisp/magit-obsolete.el
@@ -1,6 +1,6 @@
 ;;; magit-obsolete.el --- obsolete definitions  -*- lexical-binding: t -*-
 
-;; Copyright (C) 2010-2017  The Magit Project Contributors
;; Copyright (C) 2010-2018  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
 ;; lists all contributors.  If not, see http://magit.vc/authors.
@@ -121,5 +121,31 @@ use `magit-pre-refresh-hook', `magit-post-refresh-hook',
 (define-obsolete-function-alias 'magit-checkout-file
   'magit-file-checkout "Magit 2.9.1")
 
;;; Obsolete since v2.11.0

(define-obsolete-function-alias 'magit-modified-files
  'magit-unstaged-files "Magit 2.11.0")

(define-obsolete-function-alias 'magit-insert-submodules
  'magit-insert-modules-overview "Magit 2.11.0")

;;; Obsolete since v2.12.0

(define-obsolete-function-alias 'magit-insert-unpulled-from-upstream-or-recent
  'magit-insert-unpulled-from-upstream "Magit 2.12.0")

(make-obsolete-variable 'magit-no-confirm-default
                        'magit-dwim-selection "Magit 2.12.0")

(make-obsolete 'magit-section-type     "use (oref ... type) instead"     "Magit 2.12.0")
(make-obsolete 'magit-section-value    "use (oref ... value) instead"    "Magit 2.12.0")
(make-obsolete 'magit-section-start    "use (oref ... start) instead"    "Magit 2.12.0")
(make-obsolete 'magit-section-content  "use (oref ... content) instead"  "Magit 2.12.0")
(make-obsolete 'magit-section-end      "use (oref ... end) instead"      "Magit 2.12.0")
(make-obsolete 'magit-section-hidden   "use (oref ... hidden) instead"   "Magit 2.12.0")
(make-obsolete 'magit-section-washer   "use (oref ... washer) instead"   "Magit 2.12.0")
(make-obsolete 'magit-section-parent   "use (oref ... parent) instead"   "Magit 2.12.0")
(make-obsolete 'magit-section-children "use (oref ... children) instead" "Magit 2.12.0")

 (provide 'magit-obsolete)
 ;;; magit-obsolete.el ends here
