diff --git a/lisp/magit-subtree.el b/lisp/magit-subtree.el
index 494ff7c..b31d0d8 100644
--- a/lisp/magit-subtree.el
++ b/lisp/magit-subtree.el
@@ -1,6 +1,6 @@
 ;;; magit-subtree.el --- subtree support for Magit  -*- lexical-binding: t -*-
 
-;; Copyright (C) 2011-2017  The Magit Project Contributors
;; Copyright (C) 2011-2018  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
 ;; lists all contributors.  If not, see http://magit.vc/authors.
@@ -88,13 +88,15 @@
   (magit-run-git-async "subtree" subcmd (concat "--prefix=" prefix) args))
 
 ;;;###autoload
-(defun magit-subtree-add (prefix repository commit args)
-  "Add COMMIT from REPOSITORY as a new subtree at PREFIX."
-  (interactive (list (magit-subtree-prefix "Add subtree")
-                     (magit-read-string-ns "Repository")
-                     (magit-read-string-ns "Commit")
-                     (magit-subtree-args)))
-  (magit-git-subtree "add" prefix args repository commit))
(defun magit-subtree-add (prefix repository ref args)
  "Add REF from REPOSITORY as a new subtree at PREFIX."
  (interactive
   (cons (magit-subtree-prefix "Add subtree")
         (let ((remote (magit-read-remote-or-url "From repository")))
           (list remote
                 (magit-read-refspec "Ref" remote)
                 (magit-subtree-args)))))
  (magit-git-subtree "add" prefix args repository ref))
 
 ;;;###autoload
 (defun magit-subtree-add-commit (prefix commit args)
@@ -113,19 +115,21 @@
   (magit-git-subtree "merge" prefix args commit))
 
 ;;;###autoload
-(defun magit-subtree-pull (prefix repository commit args)
-  "Pull COMMIT from REPOSITORY into the PREFIX subtree."
-  (interactive (list (magit-subtree-prefix "Pull into subtree")
-                     (magit-read-string-ns "From repository")
-                     (magit-read-string-ns "Commit")
-                     (magit-subtree-args)))
-  (magit-git-subtree "pull" prefix args repository commit))
(defun magit-subtree-pull (prefix repository ref args)
  "Pull REF from REPOSITORY into the PREFIX subtree."
  (interactive
   (cons (magit-subtree-prefix "Pull into subtree")
         (let ((remote (magit-read-remote-or-url "From repository")))
           (list remote
                 (magit-read-refspec "Ref" remote)
                 (magit-subtree-args)))))
  (magit-git-subtree "pull" prefix args repository ref))
 
 ;;;###autoload
 (defun magit-subtree-push (prefix repository ref args)
   "Extract the history of the subtree PREFIX and push it to REF on REPOSITORY."
   (interactive (list (magit-subtree-prefix "Push subtree")
-                     (magit-read-string-ns "To repository")
                     (magit-read-remote-or-url "To repository")
                      (magit-read-string-ns "To reference")
                      (magit-subtree-args)))
   (magit-git-subtree "push" prefix args repository ref))
