diff --git a/.travis/before_script/linux_gcc_minimal_feature_set.sh b/.travis/before_script/linux_gcc_minimal_feature_set.sh
deleted file mode 100755
index f0f62269..00000000
-- a/.travis/before_script/linux_gcc_minimal_feature_set.sh
+++ /dev/null
@@ -1,14 +0,0 @@
#!/usr/bin/env bash

cmake -H. -B${BUILD_DIR} \
    -DCMAKE_BUILD_TYPE=Debug \
    -DUJIT_HAS_JIT=OFF \
    -DUJIT_HAS_FFI=OFF \
    -DUJIT_LUA52COMPAT=OFF \
    -DUJIT_ENABLE_GDBJIT=OFF \
    -DUJIT_ENABLE_PROFILER=OFF \
    -DUJIT_ENABLE_IPROF=OFF \
    -DUJIT_ENABLE_COVERAGE=OFF \
    -DUJIT_ENABLE_CO_TIMEOUT=OFF \
    -DUJIT_ENABLE_MEMPROF=OFF \
    -DUJIT_ENABLE_THREAD_SAFETY=OFF
