diff --git a/config/database.travis.yml b/config/database.travis.yml
deleted file mode 100644
index c6bc8da91..000000000
-- a/config/database.travis.yml
+++ /dev/null
@@ -1,12 +0,0 @@
---
development: &development
  encoding: utf8
  adapter: mysql2
  username: travis
  database: lara_dev
test: &test
  encoding: utf8
  adapter: mysql2
  username: travis
  database: lara_test
production: *development
