diff --git a/.travis.rosinstall b/.ci.rosinstall
similarity index 79%
rename from .travis.rosinstall
rename to .ci.rosinstall
index 701a644..52ac22e 100644
-- a/.travis.rosinstall
+++ b/.ci.rosinstall
@@ -1,12 +1,16 @@
 - git:
     local-name: robotont/robotont_driver
     uri: https://github.com/robotont/robotont_driver.git
+    version: 'noetic-devel'
 - git:
     local-name: robotont/robotont_msgs
     uri: https://github.com/robotont/robotont_msgs.git
+    version: 'noetic-devel'
 - git:
     local-name: robotont/robotont_description
     uri: https://github.com/robotont/robotont_description.git
+    version: 'noetic-devel'
 - git:
     local-name: robotont/robotont_nuc_description
     uri: https://github.com/robotont/robotont_nuc_description.git
+    version: 'noetic-devel'
