diff --git a/.travis/install.sh b/.travis/install.sh
deleted file mode 100755
index 562e40b..0000000
-- a/.travis/install.sh
+++ /dev/null
@@ -1,21 +0,0 @@
#!/bin/bash

scripts=/var/scripts
repo=${1}
deployment=${2}
type=${3}

source ${scripts}/${repo}/env/${type}/.env

echo "installing ${repo} as ${deployment}"
cd ${LOCATION}/${deployment}
pwd

echo "yarn install"
yarn install
echo "copy .env"
cp ${scripts}/${repo}/dotenv/${type}/${deployment} .env
echo "yarn webpack:prod"
yarn webpack:prod

echo "completed"
