diff --git a/.travis.yml b/.travis.yml
deleted file mode 100644
index 55ce3d0..0000000
-- a/.travis.yml
+++ /dev/null
@@ -1,21 +0,0 @@
matrix:
  include:
    - language: cpp
      addons:
        apt:
          packages:
            - libxerces-c-dev
            - valgrind
            
      script:
        - mkdir build
        - cd build
        - cmake ..
        - cmake --build .
        - ctest .
        - ctest -T memcheck .
        - sudo make install
      
    - language: java
      jdk:
        - openjdk8
