diff --git a/lisp/magit-margin.el b/lisp/magit-margin.el
index 5f2bef67..552839e9 100644
--- a/lisp/magit-margin.el
++ b/lisp/magit-margin.el
@@ -1,6 +1,6 @@
 ;;; magit-margin.el --- margins in Magit buffers  -*- lexical-binding: t -*-
 
-;; Copyright (C) 2010-2017  The Magit Project Contributors
;; Copyright (C) 2010-2018  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
 ;; lists all contributors.  If not, see http://magit.vc/authors.
