diff --git a/test/handler/test_gitlint_handler.vader b/test/handler/test_gitlint_handler.vader
index 60d632a..5c53166 100644
--- a/test/handler/test_gitlint_handler.vader
++ b/test/handler/test_gitlint_handler.vader
@@ -61,6 +61,19 @@ Execute(Disabling trailing whitespace warnings should work):
   \   '8: T2 Trailing whitespace',
   \])
 
  AssertEqual
  \ [
  \   {
  \     'lnum': 8,
  \     'type': 'E',
  \     'text': 'Trailing whitespace',
  \     'code': 'B2',
  \   },
  \ ],
  \ ale_linters#gitcommit#gitlint#Handle(bufnr(''), [
  \   '8: B2 Trailing whitespace',
  \])

   let b:ale_warn_about_trailing_whitespace = 0
 
   AssertEqual
@@ -68,3 +81,9 @@ Execute(Disabling trailing whitespace warnings should work):
   \ ale_linters#gitcommit#gitlint#Handle(bufnr(''), [
   \   '8: T2 Trailing whitespace',
   \ ])

  AssertEqual
  \ [],
  \ ale_linters#gitcommit#gitlint#Handle(bufnr(''), [
  \   '8: B2 Trailing whitespace',
  \ ])
